### PR TITLE
Add shuffle functionality

### PR DESCRIPTION
**Pull request by malucky**
@deini @alexmckenley

### CODE REVIEW

File: public/javascript/_common/services/playerSetting/playerSetting.service.js [hunk 1]
```diff
@@ -0,0 +1,49 @@
+(function() {
+  'use strict';
+
+  angular
+    .module('ammo.playerSetting.service', [
```
Comment (alexmckenley) [inline public/javascript/_common/services/playerSetting/playerSetting.service.js]: I feel like this should really be plural `playerSettings` because it will keep track of more than one setting.

File: public/javascript/_common/services/currentPlaylist/currentPlaylist.service.js [hunk 2]
```diff
@@ -2,29 +2,46 @@
...
+        service,
+        shuffledPlaylist,
+        unshuffledPlaylist;
 
     service = {
       setPlaylist: setPlaylist,
-      nextSong: nextSong
+      nextSong: nextSong,
+      setShuffle: setShuffle,
+      unsetShuffle: unsetShuffle
```
Comment (alexmckenley) [inline public/javascript/_common/services/currentPlaylist/currentPlaylist.service.js]: This could probably be a single function called `shuffle` that takes a single boolean param :tropical_drink:

File: public/javascript/_common/services/currentPlaylist/currentPlaylist.service.js [hunk 3]
```diff
@@ -51,5 +68,30 @@
       var songs = currentPlaylist.songs;
       return songs[index];
     }
+
+    function setShuffle() {
+      shuffledPlaylist = _.clone(playlist, true);
```
Comment (alexmckenley) [inline public/javascript/_common/services/currentPlaylist/currentPlaylist.service.js]: Looks like playlist is undefined here, you probably want unshuffledPlaylist

File: public/javascript/_common/services/currentPlaylist/currentPlaylist.service.js [hunk 4]
```diff
@@ -51,5 +68,30 @@
...
+      return currentPlaylist;
+    }
+
+    function moveCurrentSongToFirst() {
+      var currentSongs  = currentPlaylist.songs,
+          shuffledSongs = shuffledPlaylist.songs,
+          currentSongIndex = nextIndex === 0 ? nextIndex : nextIndex - 1,
+          currentShuffledSongIndex = _.indexOf(shuffledSongs, currentSongs[currentSongIndex]);
+
+      shuffledSongs.unshift(shuffledSongs.splice(currentShuffledSongIndex, 1)[0]);
```
Comment (alexmckenley) [inline public/javascript/_common/services/currentPlaylist/currentPlaylist.service.js:103]: I really think it would be worth using _.find instead of messing around with indexes all over the place (even inside of the the getNextSong function). But we could do that later as well, if all this logic currently works.

File: public/javascript/_common/services/player/player.service.js [hunk 5]
```diff
@@ -14,7 +19,8 @@
         play: play,
         pause: pause,
         unpause: unpause,
-        nextSong: nextSong
+        nextSong: nextSong,
+        toggleShuffle: toggleShuffle
```
Comment (alexmckenley) [inline public/javascript/_common/services/player/player.service.js:23]: This is fine for now, but as soon as we have the UI for it, we should delete this altogether since the playlist service does not need to know about shuffling at all.

File: public/javascript/_common/services/playerSetting/playerSetting.service.js [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+            setRepeat: setRepeat
+          };
+
+      return playerSetting;
+
+      function isShuffled() {
+        return shuffled;
+      }
+
+      function getRepeat() {
```
Comment (alexmckenley) [inline public/javascript/_common/services/playerSetting/playerSetting.service.js]: Inconsistent method names: `isShuffled` and `getRepeat` (although I'm not sure of the best solution for this). Maybe a single method called `getSetting(settingName)`? (also, maybe the same thing for `setSetting`?

File: public/javascript/_common/services/playerSetting/playerSetting.service.js [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+        return repeat;
+      }
+
+      function toggleShuffle() {
+        shuffled = !shuffled;
+        if (shuffled) {
+          event.publish('setShuffle');
+        } else {
+          event.publish('unsetShuffle');
```
Comment (alexmckenley) [inline public/javascript/_common/services/playerSetting/playerSetting.service.js]: We should probably just have a single event called 'shuffle' and we can just pass true or false as part of the event data like so:

```
event.publish('shuffle', true);
event.publish('shuffle', false);
```

File: karma.conf.js [hunk 8]
```diff
@@ -18,8 +18,7 @@ module.exports = function(config) {
       'build/js/vendor.js',
       'bower_components/angular-mocks/angular-mocks.js',
       'build/js/templates.js',
-      'build/js/bundle.js',
-      'public/**/*.spec.js'
+      'public/**/*.js'
```
Comment (alexmckenley) [inline karma.conf.js:21]: Do we need to filter out .scenario files?

Comment (malucky) [inline karma.conf.js:21]: good call. added

File: public/javascript/_common/services/currentPlaylist/currentPlaylist.service.js (hunk 4 above)
Comment (malucky) [inline public/javascript/_common/services/currentPlaylist/currentPlaylist.service.js:103]: will do in separate pr

File: public/javascript/_common/services/player/player.service.js (hunk 5 above)
Comment (malucky) [inline public/javascript/_common/services/player/player.service.js:23]: sure

**Comment by malucky** [comment]
@alexmckenley did suggested changes

**Comment by deini** [comment]
@alexmckenley Re-pinging

**Comment by alexmckenley** [comment]
:+1:  Looks good, thanks for doing this ray!
